go/samples/pgvector/main.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This program shows how to use Postgres's pgvector extension with Genkit.
+
+// This program can be manually tested like so:
+//
+// In development mode (with the environment variable GENKIT_ENV="dev"):
+// Start the server listening on port 3100:
+//
+//	go run . -dbconn "$DBCONN" -apikey $API_KEY &
+//
+// Ask a question:
+//
+//	curl -d '{"Show": "Best Friends", "Question": "Who does Alice love?"}' http://localhost:3400/askQuestion
+package main
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/firebase/genkit/go/genkit"
+	"github.com/firebase/genkit/go/plugins/googleai"
+	_ "github.com/lib/pq"
+	pgv "github.com/pgvector/pgvector-go"
+)
+
+var (
+	connString = flag.String("dbconn", "", "database connection string")
+	apiKey     = flag.String("apikey", "", "Gemini API key")
+	index      = flag.Bool("index", false, "index the existing data")
+)
+
+func main() {
+	flag.Parse()
+	if err := run(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run() error {
+	if *connString == "" {
+		return errors.New("need -dbconn")
+	}
+	if *apiKey == "" {
+		return errors.New("need -apikey")
+	}
+	ctx := context.Background()
+	_, ems, err := googleai.Init(ctx, googleai.Config{
+		APIKey:    *apiKey,
+		Embedders: []string{"embedding-001"},
+	})
+	if err != nil {
+		return err
+	}
+	embedder := ems[0]
+
+	db, err := sql.Open("postgres", *connString)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	if *index {
+		indexer := defineIndexer(db, embedder)
+		if err := indexExistingRows(ctx, db, indexer); err != nil {
+			return err
+		}
+	}
+
+	retriever := defineRetriever(db, embedder)
+
+	type input struct {
+		Question string
+		Show     string
+	}
+
+	genkit.DefineFlow("askQuestion", func(ctx context.Context, in input, _ genkit.NoStream) (string, error) {
+		res, err := ai.Retrieve(ctx, retriever, &ai.RetrieverRequest{
+			Document: &ai.Document{Content: []*ai.Part{ai.NewTextPart(in.Question)}},
+			Options:  in.Show,
+		})
+		if err != nil {
+			return "", err
+		}
+		for _, doc := range res.Documents {
+			fmt.Printf("%+v %q\n", doc.Metadata, doc.Content[0].Text)
+		}
+		// Use documents in RAG prompts.
```

Comment:
Is this a TODO? I'm not sure what the point of retrieving here is, as you don't actually answer the question asked.

Reply:
It's deliberately a sketch. So yeah it's a TODO, but not for us, for whoever copies the sample.

I based it off of https://github.com/firebase/genkit/blob/main/docs/templates/pgvector.md.
I decided to write actual code rather than a markdown file so I could be sure it actually worked.